ARM-CI: Add CI checks for Linux ARM Emulator

prajwal-aithal opened the pull request:
Some changes to the native part of corefx breaks builds for the Linux ARM Emulator. Thus adding CI checks for native part of corefx using cross build based on the emulator rootfs.

The check has been named as `Innerloop LinuxARMEmulator <config> Cross Build`. Though it is a part of the innerloop check, the code for this check has been separated from the other innerloop checks as the build steps and the archived contents are different compared to the other innerloop checks. Also, this check only builds the native binaries as at this stage we are only concerned about checking for regressions in building native component of corefx. More steps will be added as and when we make progress in the ARM-based builds for corefx.
 
Signed-off-by: Prajwal A N <an.prajwal@samsung.com>

prajwal-aithal commented:
@dotnet-bot test CI please

prajwal-aithal commented:
@jashook @Priya91 Could you please review this PR?

However, please do not merge this as there is a minor modification needed to be made to the emulator rootfs on the CI systems.

cc: @myungjoo @leemgs

stephentoub commented:
cc: @ellismg, @mmitche, @Priya91

stephentoub commented:
Test Innerloop Ubuntu14.04 Debug Build and Test please

Priya91 commented on netci.groovy:
```diff
@@ -349,4 +350,47 @@ def osShortName = ['Windows 10': 'win10',
+    ['Debug', 'Release'].each { configurationGroup ->
+        ['LinuxARMEmulator'].each { os ->
+            def osGroup = osGroupMap[os]
+            def newJobName = "${os.toLowerCase()}_${configurationGroup.toLowerCase()}"
```
this job name has to be different, the innerloop jobs have similar names, it will override the innerloop definitions..

Priya91 commented on netci.groovy:
```diff
@@ -349,4 +350,47 @@ def osShortName = ['Windows 10': 'win10',
+            // The cross build jobs run on Ubuntu. The arm-cross-latest version
+            // contains the packages needed for cross building corefx
+            Utilities.setMachineAffinity(newJob, 'Ubuntu', 'arm-cross-latest')
+
```
Ubuntu14.04.. The Ubuntu in Utilities.groovy exists to just handle the case when we were only using 14.04 and had used osname as Ubuntu.. Going forward we will keep it specific as 14.04 or 16.04..

jashook replied:
@Priya91 when do we expect to make this change? We already have infrastructure in coreclr using this machine affinity.

Priya91 replied:
> when do we expect to make this change?

This change is already made. This machine affinity will work, it just exists to not break jobs that uses Ubuntu from old naming scheme. Ubuntu14.04 or Ubuntu16.04, these are the newly defined tags, to denote machine affinity for ubuntu.

jashook replied:
Ah I see, that makes sense.

jashook commented on netci.groovy:
```diff
@@ -349,4 +350,47 @@ def osShortName = ['Windows 10': 'win10',
+            Utilities.addArchival(newJob, archiveContents)
+
+            // Set up triggers
+            Utilities.addGithubPRTriggerForBranch(newJob, branch, "Innerloop ${os} ${configurationGroup} Cross Build", "(?i).*test\\W+Innerloop\\W+Linux\\W+ARM\\W+Emulator\\W+${configurationGroup}.*")
```
It is probably worth it to add emulator to the pr trigger similar to the coreclr netci file.

prajwal-aithal replied:
I did not completely understand what you meant by this. Did you mean the job name? The os variable in the case of emulator is LinuxARMEmulator. So I think the job name would be something like `Innerloop LinuxARMEmulator <configuration> Cross Build`.

prajwal-aithal replied:
@jashook I too think it is better to have a custom name for the emulator job (as in coreclr) than the current name. I have made this change in the latest version of the PR.

prajwal-aithal commented:
@dotnet-bot test CI please

prajwal-aithal commented:
@Priya91 @jashook Could you please review the latest version of the PR (I have incorporated your previous review in the latest PR)?

Priya91 commented:
LGTM

prajwal-aithal commented:
@jashook Could you please confirm if the `corefx_cross.patch` has been downloaded to the target location (`/opt/linux-arm-emulator-root/dotnet/setenv/corefx_cross.patch`)?

prajwal-aithal commented:
@dotnet-bot test  Innerloop Windows_NT Debug Build and Test

prajwal-aithal commented:
We will have to wait for #9740 to be merged as that change fixes the error in cross building for release configuration.

prajwal-aithal commented:
@dotnet-bot Innerloop OSX Debug Build and Test

prajwal-aithal commented:
@dotnet-bot test Innerloop OSX Debug Build and Test

prajwal-aithal commented:
@jashook @Priya91 I have uploaded a new version of this PR. The changes from the previous version are:
* Added verbose to script (called in the groovy shell commands)
* Re-organized the code into functions and documented the script
* Rebased over latest master

Could you please review this?

Also, if @jashook can confirm that the emulator on the CI system has `corefx_cross.patch` at `/opt/linux-arm-emulator-root/dotnet/setenv/`, then we can merge this patch and test if the check works as expected (I will add another PR to make this check automatic on a per PR basis after confirming that the check works).

prajwal-aithal commented:
@dotnet-bot test CI please

prajwal-aithal commented:
@jashook @Priya91 Any progress on this?

prajwal-aithal commented:
@dotnet-bot test CI please

prajwal-aithal commented:
The failure in `Innerloop Ubuntu 14.04 Debug Build and Test` doesnt seem to be caused by this PR. Also, the error message in that check is not clear about the cause of failure (expected and actual arguments to the failing test are the same and yet the test fails at `Assert.Equal()`).

jashook commented:
@prajwal-aithal the corefx patch is not present on the machine and needs to be added.

prajwal-aithal commented:
@jashook I had sent a mail regarding the same last week. I have resent the mail today. Could you please check your inbox for this mail?

Priya91 commented:
@prajwal-aithal Added the patch file to the ubuntu arm image.

leemgs commented:
/CC: @sjsinju , FYI
